### PR TITLE
Add tests and dep bump

### PR DESCRIPTION
**Pull request by albertodvp**
In this PR:
- refactored test suite
- bump plutarch/plutus core versions
- add tests (nominal cases and attacks)
- moved cabal.project in the root directory to make interaction with `cabal` easier

### CODE REVIEW

File: ytxp-plutarch/test/Cardano/YTxP/Test/Control/Yielding/Scripts/Attacks.hs [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+    --   fromCustomRedeemer :: a -> Redeemer
+    --   fromCustomRedeemer = Redeemer . toBuiltinData
+    --
+    -- Does it add values? We could rewrite TxFCEKInput in term of thatYieldingRedeemer
+    -- and we could derive anyclass it. Is it safe?
+    redeemer = unsafeFromBuiltinData (toBuiltinData r)
+    (badRedeemer, sc') = appEndo attack (redeemer, sc)
+    badLedgerRedeemer = toLedgerRedeemer badRedeemer
+   in
+    Right (TxFCEKInput md badLedgerRedeemer sc' script)
```
Comment (albertodvp) [inline ytxp-plutarch/test/Cardano/YTxP/Test/Control/Yielding/Scripts/Attacks.hs:181]: I'd need some feedback for this

File: ytxp-plutarch/test/Cardano/YTxP/Test/Control/Yielding/Helper.hs
Comment (peter-mlabs) [inline ytxp-plutarch/test/Cardano/YTxP/Test/Control/Yielding/Helper.hs:1]: Empty file, can be removed

File: ytxp-plutarch/test/Cardano/YTxP/Test/Control/Yielding/Scripts.hs [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  ),
+ )
+import Control.Monad.Reader (Reader, runReader)
+import Test.Tasty (TestTree, testGroup)
+
+dummyParams :: ScriptsTestsParams
+dummyParams =
+  ScriptsTestsParams
+    { authorisedScriptHash = "22222222222222222222222222222222222222222222222222222222"
+    , authorisedScriptsSTCS = AuthorisedScriptsSTCS "bb"
```
Comment (peter-mlabs) [inline ytxp-plutarch/test/Cardano/YTxP/Test/Control/Yielding/Scripts.hs:24]: note that you won't get a run-time error when `show`ing this, but its still not well-formed. It would need to be updated if we get full phase-1 validation wired in to PCB

File: ytxp-plutarch/test/Cardano/YTxP/Test/Control/Yielding/Scripts/Attacks.hs (hunk 1 above)
Comment (peter-mlabs) [reply ytxp-plutarch/test/Cardano/YTxP/Test/Control/Yielding/Scripts/Attacks.hs:180]: Hm, so you'd basically like to just remove the wrapping and unwraping on lines 176-178? 
It sounds like we might actually just want a functor instance somewhere...

File: ytxp-plutarch/test/Cardano/YTxP/Test/Control/Yielding/Scripts/Attacks.hs [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+    (const 42)
+
+attackAuthorisedProofIndexInvalidIndex ::
+  Reader ScriptsTestsParams (Endo (YieldingRedeemer, ScriptContext))
+attackAuthorisedProofIndexInvalidIndex =
+  redeemerAttack
+    (O.authorisedScriptProofIndex . _Wrapped . _2)
+    -- NOTE: this is a big number that will be not present in
+    -- any script context created for testing purposes
+    (const 42)
```
Comment (peter-mlabs) [inline ytxp-plutarch/test/Cardano/YTxP/Test/Control/Yielding/Scripts/Attacks.hs:256]: A more stable way to do this (which would survive property testing, if requested) is to just take length (don't even need to add 1 because it's 0 indexed)

**Review by peter-mlabs — COMMENTED** [review]
I'm happy with this overall. Great work Alberto! Really appreciate your documentation, makes it very easy to review and understand what you're trying to do

File: cabal.project [hunk 4]
```diff
@@ -12,16 +12,16 @@ repository cardano-haskell-packages
...
       .
-      plutarch-api
       plutarch-extra
+      plutarch-ledger-api
 
 source-repository-package
     type: git
     location: https://github.com/input-output-hk/xsy-liqwid-libs
-    tag: 01dcaf60ee5386f0975aed908d6f92ab0e9b34e0
+    tag: 2425031da690546dd64669975598a99d1279eb16
```
Comment (the-headless-ghost) [inline cabal.project:24]: Depending on which PR we merge first, this could potentially revert the change on #21. So, I would change it to `f46485d8796db37a5b3e7fe90a830355504f4edb` instead.

File: hie.yaml [hunk 5]
```diff
@@ -0,0 +1,13 @@
+cradle:
+  cabal:
```
Comment (the-headless-ghost) [inline hie.yaml:2]: Can't we leave this to the default behavior by leaving the cradle cabal entry empty?

Comment (albertodvp) [reply hie.yaml:2]: For some project configuration this does not work as expected but it does for out setup so I'll follow your suggestion, thanks!

File: ytxp-plutarch/test/Cardano/YTxP/Test/Control/Yielding/Scripts/Attacks.hs (hunk 1 above)
Comment (albertodvp) [reply ytxp-plutarch/test/Cardano/YTxP/Test/Control/Yielding/Scripts/Attacks.hs:180]: Yeah, we should add that on sc-tools but I don't think it's valuable at this point (given the other priorities we have) so I'll leave this as it

File: ytxp-plutarch/test/Cardano/YTxP/Test/Control/Yielding/Scripts/NominalCases.hs [hunk 6]
```diff
@@ -0,0 +1,86 @@
+module Cardano.YTxP.Test.Control.Yielding.Scripts.NominalCases (
+  testNominalCasesR,
+  mintNominalCaseBuilderR,
+  spendNominalCaseBuilderR,
+  rewardNominalCaseBuilderR,
+) where
+
+import Cardano.YTxP.SDK.Redeemers (AuthorisedScriptIndex (AuthorisedScriptIndex), AuthorisedScriptProofIndex (AuthorisedScriptProofIndex), AuthorisedScriptPurpose (Minting, Rewarding, Spending), YieldingRedeemer (YieldingRedeemer))
```
[the remark this answers is not in the document]
Comment (albertodvp) [reply ytxp-plutarch/test/Cardano/YTxP/Test/Control/Yielding/Scripts/NominalCases.hs]: I've just tried that here:
https://fourmolu.github.io

and it seems fourmolu isn't splitting that :(
I've fixed that by hand

File: ytxp-plutarch/test/Cardano/YTxP/Test/Control/Yielding/Scripts/NominalCases.hs
[the remark this answers is not in the document]
Comment (albertodvp) [reply ytxp-plutarch/test/Cardano/YTxP/Test/Control/Yielding/Scripts/NominalCases.hs:1]: I've just added an input and verified the test doesn't pass bot now I have like 2 inputs but one could argue that I'm just checking "the last one", so I've moved on and put the interesting data in the middle of the list but the problem persists: even if I put the correct token at index n, I'd have to know n in order to write the correct redeemer so the test is white-boxed, I don't think doing any of these change improves the robustness of this test, we'd need property based test for this. Do you think having a less trivial context (e.g. 2 inputs/ minted tokens/...) is better than the current approach?

File: ytxp-plutarch/src/Cardano/YTxP/Control/Yielding/Scripts.hs [hunk 7]
```diff
@@ -35,13 +44,13 @@ scripts conf =
 -- Plutarch level terms
 
 -- | Yielding Validator
-yieldingValidator ::
```
Comment (albertodvp) [inline ytxp-plutarch/src/Cardano/YTxP/Control/Yielding/Scripts.hs:38]: @the-headless-ghost : do you see any problem with this?

File: docs/flows/staking-validator-management-flow/stake-validator-deregistering.md [hunk 8]
```diff
@@ -9,7 +9,7 @@
 7. [Semantics](#orge82b809)
 8. [Constraints](#org187c6f6)
 
-When a yielding or yielded-to staking validator is no longer needed, we can de-register it by submitting a transaction with a `DCertDelegDeRegKey` in the `txInfoDCert` field.
+When a yielding validator or authorised staking validator is no longer needed, we can de-register it by submitting a transaction with a `DCertDelegDeRegKey` in the `txInfoDCert` field.
```
Comment (peter-mlabs) [inline docs/flows/staking-validator-management-flow/stake-validator-deregistering.md:12]:
```suggestion
When a yielding staking validator or authorised staking validator is no longer needed, we can de-register it by submitting a transaction with a `DCertDelegDeRegKey` in the `txInfoDCert` field.
```
